### PR TITLE
8294397: Replace StringBuffer with StringBuilder within java.text

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Problem: Unnecessary instances of StringBuffer within java.text (internal only)

Fix: StringBuffer Replaced with StringBuilder, and adjusted variable/method names

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294397](https://bugs.openjdk.org/browse/JDK-8294397): Replace StringBuffer with StringBuilder within java.text

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [442c2a4e](https://git.openjdk.org/jdk/pull/10475/files/442c2a4e0a001e7de100d0d8aaba66b31b19ac35)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [442c2a4e](https://git.openjdk.org/jdk/pull/10475/files/442c2a4e0a001e7de100d0d8aaba66b31b19ac35)
 * [Brent Christian](https://openjdk.org/census#bchristi) (@bchristi-git - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10475/head:pull/10475` \
`$ git checkout pull/10475`

Update a local copy of the PR: \
`$ git checkout pull/10475` \
`$ git pull https://git.openjdk.org/jdk pull/10475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10475`

View PR using the GUI difftool: \
`$ git pr show -t 10475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10475.diff">https://git.openjdk.org/jdk/pull/10475.diff</a>

</details>

### CODE REVIEW

**Review by LanceAndersen — APPROVED** [review]
Looks good.

should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?

**Comment by justin-curtis-lu** [comment]
> should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?

Brent actually pointed that out to me as well. I believe since Bug4170614Test.java is managed by Bug4170614TestRun.java which has tags, then it would be fine to not include the tags(in Bug4170614Test.java), but please let me know your thoughts. @LanceAndersen

**Comment by bchristi-git** [comment]
> should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?

```
/*
    (this test doesn't have an at-test tag because it's run by Bug4170614TestRun.java
     instead of directly by the test harness)
*/
```
:)

**Comment by LanceAndersen** [comment]
> > should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?
> 
> ```
> /*
>     (this test doesn't have an at-test tag because it's run by Bug4170614TestRun.java
>      instead of directly by the test harness)
> */
> ```
> 
> :)

Justin, Brent

Thanks for pointing that out.   Do either of you know why this is configured as it is as I am not sure I see a reason for the wrapper?

**Comment by naotoj** [comment]
btw I find a typo there:
```
* Bug #4170614 complained that we had two iternal classes that
```

**Comment by LanceAndersen** [comment]
> AFAICT, the reason is described in the test class description, i.e., tweak to be in `java.text` package so that it can test the internal `hashCode` method.

Understand, I guess I don't see the difference from moving the tags to the actual test but there could be a subtly I am missing (wouldn't be the first time ;-) )

**Review by LanceAndersen — APPROVED** [review]

**Comment by LanceAndersen** [comment]
> > should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?
> 
> Brent actually pointed that out to me as well. I believe since Bug4170614Test.java is managed by Bug4170614TestRun.java which has tags, then it would be fine to not include the tags(in Bug4170614Test.java), but please let me know your thoughts. @LanceAndersen



> > should we update test/jdk/sun/text/IntHashtable/patch-src/java.base/java/text/Bug4170614Test.java to include `@test, @bug, @run, @summary`?
> 
> Brent actually pointed that out to me as well. I believe since Bug4170614Test.java is managed by Bug4170614TestRun.java which has tags, then it would be fine to not include the tags(in Bug4170614Test.java), but please let me know your thoughts. @LanceAndersen

Hi Justin,

Yes if the wrapper class is needed then we do not need the tags in the actual test.

File: src/java.base/share/classes/java/text/DigitList.java [hunk 1]
```diff
@@ -788,23 +788,23 @@ public String toString() {
...
+    private StringBuilder getStringBuilder() {
+        if (tempBuilder == null) {
+            tempBuilder = new StringBuilder(MAX_COUNT);
         } else {
-            tempBuffer.setLength(0);
+            tempBuilder.setLength(0);
         }
-        return tempBuffer;
+        return tempBuilder;
     }
```
Comment (bchristi-git) [inline src/java.base/share/classes/java/text/DigitList.java:808]: Is it worth considering whether it's still necessary to cache and reuse a single String[Buffer|Builder] instance when using 2022 garbage collectors?

File: src/java.base/share/classes/java/text/PatternEntry.java [hunk 2]
```diff
@@ -111,7 +111,7 @@ final String getChars() {
 
     // ===== privates =====
 
-    void addToBuffer(StringBuilder toAddTo,
+    void addToBuilder(StringBuilder toAddTo,
                      boolean showExtension,
                      boolean showWhiteSpace,
                      PatternEntry lastEntry)
```
Comment (bchristi-git) [inline src/java.base/share/classes/java/text/PatternEntry.java:117]: Nit: update the indentation on these lines

**Comment by naotoj** [comment]
> Understand, I guess I don't see the difference from moving the tags to the actual test but there could be a subtlety I am missing (wouldn't be the first time ;-) )

The runner has a @library/@build tag to build the test case as a library for the said reason.

File: src/java.base/share/classes/java/text/PatternEntry.java [hunk 3]
```diff
@@ -287,8 +287,8 @@ public PatternEntry next() throws ParseException {
         }
 
         // We re-use these objects in order to improve performance
-        private StringBuffer newChars = new StringBuffer();
-        private StringBuffer newExtension = new StringBuffer();
+        private StringBuilder newChars = new StringBuilder();
+        private StringBuilder newExtension = new StringBuilder();
```
Comment (bchristi-git) [inline src/java.base/share/classes/java/text/PatternEntry.java:291]: Again, in 2022, I don't know that this cache+reuse pattern is needed for performance.

**Comment by LanceAndersen** [comment]
> > Understand, I guess I don't see the difference from moving the tags to the actual test but there could be a subtlety I am missing (wouldn't be the first time ;-) )
> 
> The runner has a @library/@build tag to build the test case as a library for the said reason.

Ok, thans Naoto

**Comment by naotoj** [comment]
> > > Understand, I guess I don't see the difference from moving the tags to the actual test but there could be a subtlety I am missing (wouldn't be the first time ;-) )
> > 
> > 
> > The runner has a @library/@build tag to build the test case as a library for the said reason.
> 
> Ok, thanks Naoto

You are correct, Lance. Yes, the wrapper can be removed. Sorry for the confusion.

File: src/java.base/share/classes/java/text/PatternEntry.java (hunk 3 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/text/PatternEntry.java:291]: Due to some [bootstrapping issues](https://stackoverflow.com/questions/71834059/why-invokedynamic-based-string-concatenation-is-not-available-for-javac-in-java) in `java.base` explicit String concatenation is not translated into invokedynamic, so in practice it's faster to use `StringBuilder.append()` than `str1 + str2`, see e.g. https://github.com/openjdk/jdk/pull/3903/files

Comment (bchristi-git) [reply src/java.base/share/classes/java/text/PatternEntry.java:291]: Thanks. I was not suggesting that we use string concatenation, but rather just remove the `newChar`/`newExtension` fields, and replace, e.g.:
`214             newChars.setLength(0);`
with
`214            StringBuilder newChars = new StringBuilder();`
though I'll admit that I don't have a good sense of whether that's worth doing.

Comment (RogerRiggs) [reply src/java.base/share/classes/java/text/PatternEntry.java:291]: Memory you don't allocate doesn't need to be GC'd.  I'd leave it alone and save the GC overhead.

**Review by naotoj — APPROVED** [review]
LGTM. Thanks for removing the unnecessary wrapper.

**Review by LanceAndersen — APPROVED** [review]
Looks good Justin, thank you for removing the wrapper.

**Review by bchristi-git — APPROVED** [review]
Looks good.

File: src/java.base/share/classes/java/text/DigitList.java [hunk 4]
```diff
@@ -788,23 +788,23 @@ public String toString() {
...
-        buf.append(digits, 0, count);
-        buf.append("x10^");
-        buf.append(decimalAt);
-        return buf.toString();
+        StringBuilder temp = getStringBuilder();
+        temp.append("0.");
+        temp.append(digits, 0, count);
+        temp.append("x10^");
+        temp.append(decimalAt);
+        return temp.toString();
```
Comment (jonathan-gibbons) [inline src/java.base/share/classes/java/text/DigitList.java:796]: This could use chained method calls and avoid the local variable completely.

Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DigitList.java]: Thanks for pointing that out, made the switch there and in the getDouble() method as well.

File: src/java.base/share/classes/java/text/DigitList.java [hunk 5]
```diff
@@ -161,12 +161,12 @@ public final double getDouble() {
             return 0.0;
```
Comment (bchristi-git) [inline src/java.base/share/classes/java/text/DigitList.java:161]: The doc for this method is incorrect:
```
* If (count == 0) this throws a NumberFormatException, which
* mimics Long.parseLong().
```

**Review by bchristi-git — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/text/DigitList.java (hunk 5 above)
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/text/DigitList.java:161]: I believe that relates to https://bugs.openjdk.org/browse/JDK-8170389

Which I can address in a separate PR

Comment (naotoj) [reply src/java.base/share/classes/java/text/DigitList.java:161]: Good catch! Also, I think the comment should compare it with `Double.parseDouble()`

**Review by bchristi-git — APPROVED** [review]

**Pull request closed without merging**